File: py/plugins/middleware/src/genkit/plugins/middleware/_artifacts.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""Artifacts middleware for Genkit agents."""

from __future__ import annotations

from collections.abc import Awaitable, Callable
from typing import Any

from pydantic import BaseModel, Field

from genkit._ai._model import Message
from genkit._ai._tools import define_tool
from genkit._core._model import ModelRequest, ModelResponse
from genkit._core._registry import Registry
from genkit._core._typing import Artifact, Part, Role, TextPart
from genkit.middleware import BaseMiddleware, GenerateHookParams, GenerateMiddlewareContext

_ARTIFACTS_LISTING_MARKER = 'artifacts-middleware-listing'


class ArtifactsConfig(BaseModel):
    """Options for session artifact tools and prompt injection."""

    readonly: bool = Field(
        default=False,
        description=('When true, only read_artifact is provided — the model cannot create or update artifacts.'),
    )


class _ReadArtifactInput(BaseModel):
    name: str = Field(description='The name of the artifact to read.')


class _ReadArtifactOutput(BaseModel):
    name: str = Field(description='The artifact name.')
    content: str = Field(description='The text content of the artifact.')
    found: bool = Field(description='Whether the artifact was found in the session.')


class _WriteArtifactInput(BaseModel):
    name: str = Field(description='A unique name for the artifact (e.g. a filename like "report.md").')
    content: str = Field(description='The full text content of the artifact.')


class _WriteArtifactOutput(BaseModel):
    status: str = Field(description='Confirmation that the artifact was created or updated.')


def _extract_artifact_text(artifact: Artifact) -> str:
    parts: list[str] = []
    for part in artifact.parts:
        root = part.root
        if isinstance(root, TextPart) and root.text:
            parts.append(root.text)
    return '\n'.join(parts)


def _artifact_source(artifact: Artifact) -> str | None:
    meta = artifact.metadata
    if isinstance(meta, dict):
        source = meta.get('source')
        return str(source) if source is not None else None
    return None


def _build_artifact_listing(artifacts: list[Artifact]) -> str:
    if not artifacts:
        return '<artifacts>\nNo artifacts are currently available in the session.\n</artifacts>'

    lines = [
        '<artifacts>',
        'The following artifacts are available in the session. Use the read_artifact tool to view their content.',
    ]
    for art in artifacts:
        text = _extract_artifact_text(art)
        size_hint = f' ({len(text)} chars)' if text else ''
        source = _artifact_source(art)
        source_hint = f' [from: {source}]' if source else ''
        label = art.name or '(unnamed)'
        lines.append(f'  - {label}{size_hint}{source_hint}')
    lines.append('</artifacts>')
    return '\n'.join(lines)


def _inject_artifact_listing_messages(messages: list[Message], listing: str) -> list[Message]:
    """Strip prior listing parts and append a fresh listing to the system message."""
    out = list(messages)

    for i, msg in enumerate(out):
        filtered: list[Part] = []
        for part in msg.content or []:
            root = part.root
            meta = root.metadata if isinstance(root, TextPart) else None
            if isinstance(meta, dict) and meta.get(_ARTIFACTS_LISTING_MARKER):
                continue
            filtered.append(part)
        if len(filtered) != len(msg.content or []):
            out[i] = Message(role=msg.role, content=filtered)

    listing_part = Part(
        root=TextPart(text=listing, metadata={_ARTIFACTS_LISTING_MARKER: True}),
    )

    system_idx: int | None = None
    for i, msg in enumerate(out):
        if msg.role == Role.SYSTEM:
            system_idx = i
            break

    if system_idx is not None:
        msg = out[system_idx]
        out[system_idx] = Message(
            role=Role.SYSTEM,
            content=[*msg.content, listing_part],
        )
    else:
        out.insert(0, Message(role=Role.SYSTEM, content=[listing_part]))

    return out


def _inject_artifact_listing(request: ModelRequest, listing: str) -> ModelRequest:
    new_request = request.model_copy()
    new_request.messages = _inject_artifact_listing_messages(list(request.messages), listing)
    return new_request


class Artifacts(BaseMiddleware[ArtifactsConfig]):
    """Session artifact tools plus an injected artifact listing in the system prompt."""

    def tools(self, ctx: GenerateMiddlewareContext) -> list[Any]:
        scratch = Registry()
        tools: list[Any] = []

        async def read_artifact(input: _ReadArtifactInput) -> _ReadArtifactOutput:
            session = ctx.session
            if session is None:
                return _ReadArtifactOutput(
                    name=input.name,
                    content=(
                        'Artifacts-based tools are not available, as there is no active agent '
                        'session detected. Artifacts middleware only work when passed to an agent.'
                    ),
                    found=False,
                )

            artifacts = await session.get_artifacts()
            match = next((a for a in artifacts if a.name == input.name), None)
            if match is None:
                return _ReadArtifactOutput(
                    name=input.name,
                    content=f'Artifact "{input.name}" not found.',
                    found=False,
                )

            return _ReadArtifactOutput(
                name=input.name,
                content=_extract_artifact_text(match),
                found=True,
            )

        tools.append(
            define_tool(
                scratch,
                read_artifact,
                name='read_artifact',
                description=(
                    'Reads the content of a named artifact from the session. '
                    'Use this to inspect artifacts produced by sub-agents or '
                    'previously created artifacts.'
                ),
            ).action()
        )

        if not self.config.readonly:

            async def write_artifact(input: _WriteArtifactInput) -> _WriteArtifactOutput:
                session = ctx.session
                if session is None:
                    return _WriteArtifactOutput(status='Error: no active session.')

                await session.add_artifacts(Artifact(name=input.name, parts=[Part(TextPart(text=input.content))]))
                return _WriteArtifactOutput(status=f'Artifact "{input.name}" saved successfully.')

            tools.append(
                define_tool(
                    scratch,
                    write_artifact,
                    name='write_artifact',
                    description=(
                        'Creates or updates a named artifact in the session. '
                        'If an artifact with the same name already exists, it will be '
                        'replaced. Use this to produce files, reports, code, or other '
                        'deliverables.'
                    ),
                ).action()
            )

        return tools

    async def wrap_generate(
        self,
        params: GenerateHookParams,
        ctx: GenerateMiddlewareContext,
        next_fn: Callable[[GenerateHookParams, GenerateMiddlewareContext], Awaitable[ModelResponse]],
    ) -> ModelResponse:
        session = ctx.session
        artifacts = await session.get_artifacts() if session is not None else []
        listing = _build_artifact_listing(artifacts)
        params.request = _inject_artifact_listing(params.request, listing)
        return await next_fn(params, ctx)

File: js/plugins/middleware/src/artifacts.ts
/**
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  generateMiddleware,
  ToolAction,
  z,
  type GenerateMiddleware,
} from 'genkit';
import { tool, type Artifact } from 'genkit/beta';

// ---------------------------------------------------------------------------
// Schema
// ---------------------------------------------------------------------------

export const ArtifactsOptionsSchema = z.object({
  readonly: z
    .boolean()
    .optional()
    .describe(
      'When true, only the read_artifact tool is provided — the model ' +
        'cannot create or update artifacts. Defaults to false.'
    ),
});

export type ArtifactsOptions = z.infer<typeof ArtifactsOptionsSchema>;

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Extracts the text content from an artifact's parts, joining multiple
 * text parts with newlines.
 */
function extractArtifactText(artifact: Artifact): string {
  return (artifact.parts ?? [])
    .map((p) => p.text ?? '')
    .filter((t) => t.length > 0)
    .join('\n');
}

/**
 * Builds the `<artifacts>` system prompt block listing available artifacts.
 */
function buildArtifactListing(artifacts: Artifact[]): string {
  if (artifacts.length === 0) {
    return (
      `<artifacts>\n` +
      `No artifacts are currently available in the session.\n` +
      `</artifacts>`
    );
  }

  const listing = artifacts
    .map((a) => {
      const text = extractArtifactText(a);
      const sizeHint = text.length > 0 ? ` (${text.length} chars)` : '';
      const source = a.metadata?.source ? ` [from: ${a.metadata.source}]` : '';
      return `  - ${a.name || '(unnamed)'}${sizeHint}${source}`;
    })
    .join('\n');

  return (
    `<artifacts>\n` +
    `The following artifacts are available in the session. Use the ` +
    `read_artifact tool to view their content.\n` +
    `${listing}\n` +
    `</artifacts>`
  );
}

// ---------------------------------------------------------------------------
// Middleware
// ---------------------------------------------------------------------------

/**
 * Creates a middleware that gives the model tools to interact with session
 * artifacts, and injects an artifact listing into the system prompt.
 *
 * **Tools provided:**
 *
 * - `read_artifact` — reads an artifact by name from the session and returns
 *   its text content.
 * - `write_artifact` (unless `readonly: true`) — creates or updates an
 *   artifact in the session. Artifacts are deduplicated by name: writing to
 *   an existing name replaces the artifact.
 *
 * **System prompt injection:**
 *
 * An `<artifacts>` block is injected into (or appended to) the system message
 * on each generate turn, listing the names and sizes of all artifacts
 * currently in the session. This lets the model know what's available without
 * consuming context on the full content.
 *
 * This middleware is useful standalone (e.g. for a workspace-builder agent
 * that creates files as artifacts) or combined with the `agents` middleware
 * using `artifactStrategy: 'session'`, where sub-agent artifacts are merged
 * into the parent session and the model accesses them via these tools.
 *
 * @example
 * ```typescript
 * // Standalone: agent that creates and reads artifacts
 * const builder = ai.defineAgent({
 *   name: 'builder',
 *   system: 'You are a code generator. Use write_artifact to create files.',
 *   use: [artifacts()],
 * });
 *
 * // Combined with agents middleware (session strategy)
 * const orchestrator = ai.defineAgent({
 *   name: 'orchestrator',
 *   system: 'You coordinate sub-agents and review their work.',
 *   use: [
 *     agents({
 *       agents: ['researcher', 'coder'],
 *       artifactStrategy: 'session',
 *     }),
 *     artifacts({ readonly: true }), // can read sub-agent artifacts
 *   ],
 * });
 * ```
 */
export const artifacts: GenerateMiddleware<typeof ArtifactsOptionsSchema> =
  generateMiddleware(
    {
      name: 'artifacts',
      description:
        'Provides read_artifact and write_artifact tools for interacting ' +
        'with session artifacts, and injects an artifact listing into the ' +
        'system prompt.',
      configSchema: ArtifactsOptionsSchema,
    },
    ({ config, ai }) => {
      const readonly = config?.readonly ?? false;

      // ── read_artifact tool ──────────────────────────────────────────

      const readArtifactTool = tool(
        {
          name: 'read_artifact',
          description:
            'Reads the content of a named artifact from the session. ' +
            'Use this to inspect artifacts produced by sub-agents or ' +
            'previously created artifacts.',
          inputSchema: z.object({
            name: z.string().describe('The name of the artifact to read.'),
          }),
          outputSchema: z.object({
            name: z.string().describe('The artifact name.'),
            content: z.string().describe('The text content of the artifact.'),
            found: z
              .boolean()
              .describe('Whether the artifact was found in the session.'),
          }),
        },
        async (input) => {
          // `ai.currentSession()` throws when there is no active session.
          let session: ReturnType<typeof ai.currentSession>;
          try {
            session = ai.currentSession();
          } catch {
            return {
              name: input.name,
              content: 'Error: no active session.',
              found: false,
            };
          }

          const artifacts = session.getArtifacts();
          const artifact = artifacts.find((a) => a.name === input.name);

          if (!artifact) {
            return {
              name: input.name,
              content: `Artifact "${input.name}" not found.`,
              found: false,
            };
          }

          return {
            name: input.name,
            content: extractArtifactText(artifact),
            found: true,
          };
        }
      );

      // ── write_artifact tool ─────────────────────────────────────────

      const writeArtifactTool = tool(
        {
          name: 'write_artifact',
          description:
            'Creates or updates a named artifact in the session. ' +
            'If an artifact with the same name already exists, it will be ' +
            'replaced. Use this to produce files, reports, code, or other ' +
            'deliverables.',
          inputSchema: z.object({
            name: z
              .string()
              .describe(
                'A unique name for the artifact (e.g. a filename like "report.md").'
              ),
            content: z
              .string()
              .describe('The full text content of the artifact.'),
          }),
          outputSchema: z.object({
            status: z
              .string()
              .describe(
                'Confirmation that the artifact was created or updated.'
              ),
          }),
        },
        async (input) => {
          // `ai.currentSession()` throws when there is no active session.
          let session: ReturnType<typeof ai.currentSession>;
          try {
            session = ai.currentSession();
          } catch {
            return { status: 'Error: no active session.' };
          }

          session.addArtifacts([
            {
              name: input.name,
              parts: [{ text: input.content }],
            },
          ]);

          return { status: `Artifact "${input.name}" saved successfully.` };
        }
      );

      // ── Assemble tools ──────────────────────────────────────────────

      const artifactTools: ToolAction<any, any>[] = [readArtifactTool];
      if (!readonly) {
        artifactTools.push(writeArtifactTool);
      }

      return {
        tools: artifactTools,

        generate: async (envelope, ctx, next) => {
          const { request } = envelope;

          // ── Build artifact listing for the system prompt ──────────
          // `ai.currentSession()` throws when there is no active session.
          let currentArtifacts: Artifact[] = [];
          try {
            currentArtifacts = ai.currentSession().getArtifacts();
          } catch {
            // No active session — nothing to list.
          }

          const artifactListing = buildArtifactListing(currentArtifacts);

          // ── Inject / update the listing in the system message ─────
          const messages = [...request.messages];
          const MARKER_KEY = 'artifacts-middleware-listing';

          // Remove any previously injected listing (refresh each turn).
          for (let i = 0; i < messages.length; i++) {
            const msg = messages[i];
            const filteredContent = msg.content.filter(
              (part) => !part.metadata?.[MARKER_KEY]
            );
            if (filteredContent.length !== msg.content.length) {
              messages[i] = { ...msg, content: filteredContent };
            }
          }

          // Append to existing system message or create one.
          const systemIdx = messages.findIndex((m) => m.role === 'system');
          if (systemIdx !== -1) {
            messages[systemIdx] = {
              ...messages[systemIdx],
              content: [
                ...messages[systemIdx].content,
                {
                  text: artifactListing,
                  metadata: { [MARKER_KEY]: true },
                },
              ],
            };
          } else {
            messages.unshift({
              role: 'system',
              content: [
                {
                  text: artifactListing,
                  metadata: { [MARKER_KEY]: true },
                },
              ],
            });
          }

          return next({ ...envelope, request: { ...request, messages } }, ctx);
        },
      };
    }
  );
